File: scala/a10.scala
import scala.io.StdIn._

object Main {
  def main(args: Array[String]): Unit = {
    val n = readInt()
    val a = readLine().split(" ").map(_.toInt).toVector
    val d = readInt()
    val lr = Vector.fill(d){0}.map(_ => readLine().split(" ").map(_.toInt).toVector)

    val x = Array.fill(n){0}
    val y = Array.fill(n){0}

    x(0) = a(0)
    for (i <- 1 to (n-1)){
      x(i) = Array(x(i-1), a(i)).max
    }

    y(n-1) = a(n-1)

    for (i <- (n-2) to 0 by -1) {
      y(i)= Array(y(i+1), a(i)).max
    }

    for (i <- 0 until d )
      println(Array(x(lr(i)(0)-2), y(lr(i)(1))).max)
    
  }
}

File: cpp/a10.cpp
#include <iostream>
using namespace std;
 
int n, d;
int a[100009];
int x[100009], y[100009];
int l[100009], r[100009];

int main() {
    cin >> n;
    for (int i = 0; i < n; i++)
    {
        cin >> a[i];
    }
    cin >> d;
    for (int i = 0; i < d; i++)
    {
        cin >> l[i] >> r[i];
    }
    x[0] = a[0];
    for (int i = 1; i < n; i++)
    {
        x[i] = max(x[i-1], a[i]);
    }
    y[n-1] = a[n-1];
    for (int i = n-2; i >= 0; i--)
    {
        y[i] = max(y[i+1], a[i]);
    }

    for (int i = 0; i < d; i++)
    {
        cout << max(x[l[i]-2], y[r[i]]) << endl;
    }
    

    return 0;
}